SteamKit2/SteamKit2/Steam/Handlers/SteamMatchmaking/Lobby.cs

```diff
@@ -0,0 +1,383 @@
+                }
+
+                SteamID = steamId;
+                LobbyType = lobbyType;
+                LobbyFlags = lobbyFlags;
+                OwnerSteamID = ownerSteamId;
+                metadata = metadata ?? new Dictionary<string, string>();
+                MaxMembers = maxMembers;
+                NumMembers = numMembers;
+                members = members ?? new List<Member>();
+                Distance = distance;
+                Weight = weight;
+            }
+
+            /// <summary>
+            /// Return a new <see cref="Lobby"/> instance with equivalent data as this lobby, but is safe to
+            /// access on another thread.
+            /// </summary>
+            /// <returns>A new <see cref="Lobby"/> instance that is equivalent to this one.</returns>
+            public Lobby Clone()
+            {
+                return new Lobby(
+                    SteamID,
+                    LobbyType,
+                    LobbyFlags,
+                    OwnerSteamID,
+                    new Dictionary<string, string>( metadata ),
+                    MaxMembers,
+                    NumMembers,
+                    new List<Member>( members ), // We're not doing a deep copy as Member is never mutated.
+                    Distance,
+                    Weight
+                );
+            }
+
+            internal static byte[] EncodeMetadata( Dictionary<string, string> metadata )
+            {
+                using ( var ms = new MemoryStream() )
+                {
+                    using ( var writer = new BinaryWriter( ms ) )
+                    {
+                        byte[] header = { 0, 0 };
+                        byte[] footer = { 8, 8 };
+
+                        writer.Write( header );
+
+                        foreach ( var pair in metadata )
+                        {
+                            if ( pair.Value == null ) continue;
+                            writer.Write( ( byte )1 );
+                            writer.Write( UTF8.GetBytes( pair.Key ) );
+                            writer.Write( ( byte )0 );
+                            writer.Write( UTF8.GetBytes( pair.Value ) );
+                            writer.Write( ( byte )0 );
+                        }
+
+                        ;
+
+                        writer.Write( footer );
+                    }
+
+                    return ms.ToArray();
+                }
+            }
+
+            internal static Dictionary<string, string> DecodeMetadata( byte[] data )
+            {
+                Dictionary<string, string> metadata = new Dictionary<string, string>();
+
+                bool parsingKey = true;
+                string key = null;
+
+                int dataEnd = data.Length - 3;
+
+                if ( dataEnd < 0 || data[ 0 ] != 0 || data[ 1 ] != 0 || data[ 2 ] != 1 || data[ dataEnd + 1 ] != 8 || data[ dataEnd + 1 ] != 8 )
```

Comment:
These encoder/decoder functions could do with some unit tests.

Reply:
As per below, I may be able to piggyback off existing functionality, so unit tests may not be need.

However, the lack of tests in general was concerning me. So I'm actually putting together integration tests at the moment. Granted they can't easily be run by a CI, because you need a Steam account. So I won't push them, but will link out to it in the PR. The idea is you drop in some Steam credentials and it'll run some tests.